src/NServiceBus.Core/Unicast/MessageHandlerRegistry.cs

```diff
@@ -22,28 +22,41 @@ internal MessageHandlerRegistry(Conventions conventions)
         /// Gets the list of handlers <see cref="Type" />s for the given
         /// <paramref name="messageType" />.
         /// </summary>
-        public IEnumerable<MessageHandler> GetHandlersFor(Type messageType)
+        // ReSharper disable once ReturnTypeCanBeEnumerable.Global
+        public List<MessageHandler> GetHandlersFor(Type messageType)
         {
             Guard.AgainstNull(nameof(messageType), messageType);
             if (!conventions.IsMessageType(messageType))
             {
-                return Enumerable.Empty<MessageHandler>();
+                return noMessageHandlers;
             }
 
-            return from handlersAndMessages in handlerAndMessagesHandledByHandlerCache
-                   from messagesBeingHandled in handlersAndMessages.Value
-                   where Type.GetTypeFromHandle(messagesBeingHandled.MessageType).IsAssignableFrom(messageType)
-                   select new MessageHandler(messagesBeingHandled.MethodDelegate, Type.GetTypeFromHandle(handlersAndMessages.Key));
+            var messageHandlers = new List<MessageHandler>();
+            // ReSharper disable once LoopCanBeConvertedToQuery
+            foreach (var handlersAndMessages in handlerAndMessagesHandledByHandlerCache)
+            {
+                var handlerType = handlersAndMessages.Key;
+                // ReSharper disable once LoopCanBeConvertedToQuery
+                foreach (var messagesBeingHandled in handlersAndMessages.Value)
+                {
+                    if (messagesBeingHandled.MessageType.IsAssignableFrom(messageType))
+                    {
+                        messageHandlers.Add(new MessageHandler(messagesBeingHandled.MethodDelegate, handlerType));
+                    }
+                }
+            }
+            return messageHandlers;
         }
 
         /// <summary>
         /// Lists all message type for which we have handlers.
         /// </summary>
+        /// <remarks>This method should not be called on a hot path.</remarks>
```

Comment:
This class has a dual usage. Once on the hot path and once on the bootstrapping path. Ideally we could separate these concerns. Since I didn't want to scope creep I decided to add a remark